datastream-server/src/main/java/com/linkedin/datastream/server/Coordinator.java

```diff
@@ -415,6 +444,90 @@ public void onPartitionMovement(Long notifyTimestamp) {
     _eventQueue.put(CoordinatorEvent.createPartitionMovementEvent(notifyTimestamp));
     _log.info("Coordinator::onPartitionMovement completed successfully");
   }
+
+  /**
+   * {@inheritDoc}
+   * Stop all the tasks and wait for new session to connect.
+   */
+  @Override
+  public void onSessionExpired() {
+    _log.info("Coordinator::onSessionExpired is called");
+    _zkSessionExpired = true;
+
+    if (_shutdown) {
+      return;
+    }
+    stopEventThread();
+
+    _leaderDatastreamAddOrDeleteEventScheduled.set(false);
+    if (_leaderDatastreamAddOrDeleteEventScheduledFuture != null) {
+      _leaderDatastreamAddOrDeleteEventScheduledFuture.cancel(true);
+      _leaderDatastreamAddOrDeleteEventScheduledFuture = null;
+    }
+
+    _leaderDoAssignmentScheduled.set(false);
+    if (_leaderDoAssignmentScheduledFuture != null) {
+      _leaderDoAssignmentScheduledFuture.cancel(true);
+      _leaderDoAssignmentScheduledFuture = null;
+    }
+
+    _eventQueue.clear();
+
+    // Stopping all the connectors so that they stop producing.
+    List<Future<Boolean>> assignmentChangeFutures = _connectors.keySet().stream()
+        .map(connectorType -> {
+          _assignmentChangeThreadPool.get(connectorType).shutdownNow();
+          _assignmentChangeThreadPool.put(connectorType, Executors.newSingleThreadExecutor());
+          return dispatchAssignmentChangeIfNeeded(connectorType, new ArrayList<>(), false, false);
+        })
+        .filter(Objects::nonNull)
+        .collect(Collectors.toList());
+
+    onDatastreamChange(new ArrayList<>());
+    // Shutdown the event producer to stop any further production of records.
+    // Event producer shutdown sequence does not need to wait for onAssignmentChange to complete.
+    // This will ensure that even if any task thread does not respond to thread interruption, it will
+    // still not be able to produce any records to destination.
+    for (DatastreamTask task : _assignedDatastreamTasks.values()) {
+      ((EventProducer) task.getEventProducer()).shutdown(true);
```

Comment:
Just leaving this comment here for tracking. As discussed offline, this is not sufficient as this only shutdown the producer for currently assigned tasks. Previously assigned tasks may have been unassigned from a producer and that producer will never get closed. The fix is involved so will be done as a separate PR. From our discussion it sounded like we will need to modify the interface for the TransportProvider, feel free to assign these to me, I'd be happy to help.

Problematic race condition:

> We create a pool of producers which is shared by all tasks for a given connecter -> Kafka broker mapping. There may have been a sequence of previously assigned tasks that are assigned producers that aren't assigned to any of the current tasks. Shouldn't we close all producers to be safe?
> For example, say we create 10 producers per connector -> Kafka broker mapping.
> initialize task t1 -> producer p1
> initialize task t2 -> producer p2
> initialize task t3 -> producer p3
> assignment at this point in time: t1, t2, t3, unused producers from p4-10
> uninitialize task t2 -> which will unassign t2 from producer p2's task list
> assignment at this point in time: t1, t3, unused producers from p4-10, and p2 should also be unused, but wasn't closed
> ZK session expiry occurs
> We'll call task.getEventProducer()).shutdown(true) for t1 and t3, which will close producer p1 and p3. Producer p2 may still be running (send is taking a long time) in this scenario, since t2 was only removed from the task list of p2 and the producer was not actually shutdown.